File: src/c/graphs/test_dijkstra.c
#include <stdio.h>
#include <stdlib.h>
#include "../utils/utils.h"

typedef struct WeightedEdge {
    int to;
    double weight;
    struct WEdge *next;
} WeightedEdge;

typedef struct {
    WeightedEdge **heads;
    int num_nodes;
    int num_edges;
} WeightedGraph;

WeightedGraph *weighted_graph_create(int num_nodes);
void    weighted_graph_add_edge(WeightedGraph *g, int from, int to, double weight);
void    weighted_graph_free(WeightedGraph *g);
void    dijkstra(const WeightedGraph *g, int source, double *dist, int *visited);

static WeightedGraph *build_graph(const WeightedGraphData *gd) {
    WeightedGraph *g = weighted_graph_create(gd->num_nodes);
    for (int i = 0; i < gd->num_edges; i++) {
        weighted_graph_add_edge(g, gd->from[i], gd->to[i], gd->weight[i]);
    }
    return g;
}

int main(int argc, char *argv[]) {
    const char *size = "small";
    if (argc >= 2) size = argv[1];

    char path[256];
    snprintf(path, sizeof(path), "../../../datasets/graphs_weighted/%s.bin", size);

    WeightedGraphData *gd = load_weighted_graph_data(path);
    printf("Dataset: %s\n", path);
    printf("Loaded: %d nodes, %d edges\n", gd->num_nodes, gd->num_edges);

    WeightedGraph *g = build_graph(gd);

    double *dist    = (double *)malloc(g->num_nodes * sizeof(double));
    int    *visited = (int *)malloc(g->num_nodes * sizeof(int));

    dijkstra(g, 0, dist, visited);

    /* Count reachable nodes and find max distance */
    int reachable = 0;
    double max_dist = 0.0;
    for (int i = 0; i < g->num_nodes; i++) {
        if (dist[i] < 1e18) {
            reachable++;
            if (dist[i] > max_dist) max_dist = dist[i];
        }
    }

    printf("Dijkstra from node 0:\n");
    printf("  Reachable nodes : %d / %d\n", reachable, g->num_nodes);
    printf("  Max distance    : %.2f\n", max_dist);
    printf("  dist[0]         : %.2f\n", dist[0]);
    printf("  dist[1]         : %.2f\n", dist[1]);
    printf("  dist[2]         : %.2f\n", dist[2]);
    printf("  dist[3]         : %.2f\n", dist[3]);

    free(dist);
    free(visited);
    weighted_graph_free(g);
    free_weighted_graph(gd);
    return 0;
}

File: src/js/graphs/test_dijkstra.mjs
import { readFileSync } from 'fs';
import { dijkstra } from './dijkstra.mjs';

const size = process.argv[2] || 'small';
const path = `../../../datasets/graphs_weighted/${size}.bin`;

const buffer = readFileSync(path);
const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
const numOfNodes = view.getInt32(0, true);
const numOfEdges = view.getInt32(4, true);

const from = new Int32Array(numOfEdges);
const to = new Int32Array(numOfEdges);
const weight = new Float64Array(numOfEdges);

let offset = 8;
for (let i = 0; i < numOfEdges; i++) {
    from[i] = view.getInt32(offset, true); offset += 4;
    to[i] = view.getInt32(offset, true); offset += 4;
    weight[i] = view.getFloat64(offset, true); offset += 8;
}

// CSR structure
const offsets = new Int32Array(numOfNodes + 1);
const neighbors = new Int32Array(numOfEdges);
const weights = new Float64Array(numOfEdges);
const counts = new Int32Array(numOfNodes);

for (let i = 0; i < numOfEdges; i++) {
    counts[from[i]]++;
}

for (let i = 0; i < numOfNodes; i++) {
    offsets[i + 1] = offsets[i] + counts[i];
}

const cursor = new Int32Array(numOfNodes);
for (let i = 0; i < numOfEdges; i++) {
    const u = from[i];
    const pos = offsets[u] + cursor[u]++;
    neighbors[pos] = to[i];
    weights[pos] = weight[i];
}

const graphData = { numOfNodes, numOfEdges, offsets, neighbors, weights };

const dist = new Float64Array(graphData.numOfNodes);
const visited = new Int32Array(graphData.numOfNodes);

dijkstra(graphData, 0, dist, visited);

const reachable = Array.from(dist).filter(d => d < 1e18).length;
const maxDist = Array.from(dist).filter(d => d < 1e18).reduce((a, b) => Math.max(a, b), 0);

// console.log('------------');
// console.log('Dijkstra');
// console.log('------------');
// console.log(`Dataset: ${path}`);
// console.log(`Loaded: ${numOfNodes} nodes, ${numOfEdges} edges`);
// console.log(`Dijkstra from node 0: `);
// console.log(`Reachable nodes : ${reachable} / ${numOfNodes}`);
// console.log(`Max distance    : ${maxDist.toFixed(2)}`);
// console.log(`dist[0]         : ${dist[0].toFixed(2)}`);
// console.log(`dist[1]         : ${dist[1].toFixed(2)}`);
// console.log(`dist[2]         : ${dist[2].toFixed(2)}`);
// console.log(`dist[3]         : ${dist[3].toFixed(2)}`);

console.log('Dijkstra');
console.log('RESULT', JSON.stringify({
    nodes: numOfNodes,
    reachable,
    max_dist: maxDist.toFixed(2),
    dist0: dist[0].toFixed(2),
    dist1: dist[1].toFixed(2),
    dist2: dist[2].toFixed(2),
    dist3: dist[3].toFixed(2)
}));